arangod/Cluster/ClusterMethods.cpp

```diff
@@ -3301,24 +3301,36 @@ arangodb::Result matchBackupServersSlice(VPackSlice const planServers,
   match.clear();
 
   // Local copy of my servers
-  std::unordered_set<std::string> localCopy;
+  // Note that we use a `std::set` here for the following reasons:
+  //  - this function is supposed to return a canonical result, which
+  //    does not depend on the order of the input in `planServers` or
+  //    `dbServers`, therefore we use a sorted set here and a sorted
+  //    map as the result.
+  //  - Performance does not matter at all, this method is called for
+  //    hotbackup download and hotbackup restore, both methods are
+  //    using considerable amounts of time. The server lists are usually
+  //    quite small (3 or 5 or a few dozen at most).
+  // So please do not "optimize" this.
+  std::set<std::string> localCopy;
   std::copy(dbServers.begin(), dbServers.end(),
             std::inserter(localCopy, localCopy.end()));
 
-  // Skip all direct matching names in pair and remove them from localCopy
-  std::unordered_set<std::string>::iterator it;
+  // Skip all direct matching names in pair and remove them from localCopy.
+  // If later a server does not occur it means that no translation has to
+  // happen for it.
   for (auto planned : VPackObjectIterator(planServers)) {
     auto const plannedStr = planned.key.copyString();
```

Comment:
If we already talk about relying on the order of something, does it matter how we iterate through the `planServers` object? I remember that there were some changed to the `VPackObjectIterator` lately, that affected the order. `planServers` object might change, while being semantically the same?

Reply:
IMO the order of iteration over `planServers` should not matter here.
For every string contained in it, we only remove it from `localCopy` if it is contained in `localCopy`. If it's not contained in `localCopy`, then we emplace it into `match`.
However, `localCopy` is an std::set and thus unique and ordered. The order in which we remove from it should thus not matter.
And `match` is an `std::map` and thus unique and ordered as well. The order in which we insert into it should not matter.

Reply:
This was exactly my thinking.